---
layout: shiny_app
title: Genome browser
type: genome_browser
permalink: /genome_browser
---

## Instruction

1. Click on `Add LD blocks, CUT, SUT and XUT annotations` for loading the non-coding RNA annotations. Features are seen at the chromosome level.

2. Choose your PPI of interest and your drug of interest from the choice boxes.

3. Click on `Add piQTL track (DRUG & MTX +)` and `Add piQTL track (noDRUG & MTX +)` to load the piQTLs tracks from the PPI of interest and the drug of interest that you selected in 2. 

4. Write a gene name or its associated coordinates (`Chr:Start-End`) and click on the Search button to automatically zoom on that particular region of interest. At the chromosome level, clicking on a SNP will provide information on that SNP (ID, chromosome position, statistic values from the association test -effect size, standard error, p-value) Moreover, it is also providing the genome type location and if applicable, the name of that genomic feature. 

> Note¹: Repeating step 2 with other PPIs with or without other drug conditions will load the additional piQTL tracks in the browser.

> Note²: Several options are available by clicking on the gear icon (modification of display option or deletion of track).

*The graphs are interactive.*
