{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip,re, roman \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gffpandas.gffpandas as gffpd\n",
    "\n",
    "from tqdm import tqdm\n",
    "from Bio import SeqIO\n",
    "\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_chrom(x):\n",
    "    if '_' not in x :\n",
    "        x = x.replace('CHR', 'CHR_')\n",
    "        return x\n",
    "    return x\n",
    "\n",
    "def format_roman_chrom(x):\n",
    "    if 'MT' not in x : \n",
    "        rom_nb = x.split('_')[1]\n",
    "        nb = roman.fromRoman(rom_nb)\n",
    "        x = 'CHR_'+ str(nb)\n",
    "        return x \n",
    "    return x\n",
    "\n",
    "def format_numeric_chrom(x):\n",
    "    if 'MT' not in x : \n",
    "        num_nb = x.replace('CHR','')\n",
    "        nb = roman.toRoman(int(num_nb))\n",
    "        x = 'chr'+ str(nb)\n",
    "        return x \n",
    "    return 'chrXVII'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! for fasta in /home/savvy/PROJECTS/PHD/piQTL/data/genome_annotations/sgd_database/original/*.gz ; do echo $fasta ; zcat $fasta | grep \">\" | wc -l ; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_annot_rossi_folder = '../../data/genome_annotations/rossi_et_al_2021'\n",
    "PROMOTER_TABLE = pd.read_csv(f'{genome_annot_rossi_folder}/nfr_ndr_feature_positions.csv').rename(columns={\"Chromosome_Name\":\"seq_id\", \"systematic_id\":\"locus_id\", \"common_name\":\"name\", \"NFR/NDR_Start\":\"start\", \"NFR/NDR_End\":\"end\", \"nucleosome_stability\":\"description\"})\n",
    "PROMOTER_TABLE['seq_id'] = PROMOTER_TABLE['seq_id'].apply(format_numeric_chrom)\n",
    "PROMOTER_TABLE['source'] = 'rossi_2021_promoters'\n",
    "PROMOTER_TABLE['type'] = 'promoter'\n",
    "PROMOTER_TABLE['score'] = 1\n",
    "PROMOTER_TABLE['phase'] = '.'\n",
    "PROMOTER_TABLE['name'] = PROMOTER_TABLE['name'].replace('IMP2\\'','IMP21')\n",
    "PROMOTER_TABLE['attributes'] = [f\"Name={PROMOTER_TABLE['name'][idx]};Desc={PROMOTER_TABLE['description'][idx]}\" for idx in PROMOTER_TABLE.index ]\n",
    "prom_plus = PROMOTER_TABLE[PROMOTER_TABLE['strand'] == '+'].reset_index(drop=True)\n",
    "prom_minus = PROMOTER_TABLE[PROMOTER_TABLE['strand'] == '-'].rename(columns={\"start\":\"end\", \"end\":\"start\"}).reset_index(drop=True)\n",
    "\n",
    "PROMOTER_TABLE = pd.concat([prom_plus, prom_minus]).reset_index(drop=True)\n",
    "PROMOTER_TABLE[['seq_id', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes', 'locus_id']].to_csv(f'{genome_annot_rossi_folder}/Rossi_2021_Promoters_V64.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Format SGD information table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_annot_sgd_folder = '../../data/genome_annotations/sgd_database/original'\n",
    "genome_annot_rossi_folder = '../../data/genome_annotations/rossi_et_al_2021'\n",
    "\n",
    "#### OTHER FEATURES\n",
    "others_annotations = []\n",
    "fasta_gz = f'{genome_annot_sgd_folder}/other_features_genomic_R64-3-1_20210421.fasta.gz'\n",
    "with gzip.open(fasta_gz, \"rt\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        ID, location, genome_version, others_type = record.description.split(', ')[:4]\n",
    "        others_id = ID.split(' ')[0]\n",
    "        others_name = ID.split(' ')[1]\n",
    "        sgd_id = ID.split(' ')[2]\n",
    "        res = re.findall(r'(Chr [A-Za-z]+) from ([0-9]+)-([0-9]+)', str(location))[0]\n",
    "        Chr = res[0].upper().replace(' ','_').replace('MITO', 'MT')\n",
    "        description = ' ; '.join(record.description.split(', ')[4:])\n",
    "        if 'reverse complement' == others_type:\n",
    "            Start = res[2]\n",
    "            End = res[1]\n",
    "            cds_type = description.split(' ; ')[0]\n",
    "            description = ' ; '.join(description.split(' ; ')[1:])\n",
    "        else : \n",
    "            Start = res[1]\n",
    "            End = res[2]\n",
    "        \n",
    "        others_annotations.append([others_id, others_name, sgd_id, others_type, Chr, Start, End, description, genome_version])\n",
    "\n",
    "OTHERS_TABLE = pd.DataFrame(others_annotations, columns=['locus_id', 'name', 'sgd_id', 'genome_annotations', 'chrom', 'start', 'end', 'description', 'genome_version'])\n",
    "OTHERS_TABLE['chrom'] = OTHERS_TABLE['chrom'].apply(format_roman_chrom)\n",
    "OTHERS_TABLE.to_csv('../../data/genome_annotations/sgd_database/other_features_genomic_R64-3-1.csv', index=False)\n",
    "OTHERS_dict = OTHERS_TABLE.set_index('locus_id').to_dict()\n",
    "\n",
    "\n",
    "### RNA\n",
    "rna_annotations = []\n",
    "fasta_gz = f'{genome_annot_sgd_folder}/rna_coding_R64-3-1_20210421.fasta.gz'\n",
    "with gzip.open(fasta_gz, \"rt\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        ID, location, genome_version, rna_type = record.description.split(', ')[:4]\n",
    "        rna_id = ID.split(' ')[0]\n",
    "        rna_name = ID.split(' ')[1]\n",
    "        sgd_id = ID.split(' ')[2]\n",
    "        res = re.findall(r'(Chr [A-Za-z]+) from ([0-9]+)-([0-9]+)', str(location))[0]\n",
    "        Chr = res[0].upper().replace(' ','_').replace('MITO', 'MT')\n",
    "        description = (' ; '.join(record.description.split(', ')[4:]))\n",
    "        if 'reverse complement' == rna_type:\n",
    "            Start = res[2]\n",
    "            End = res[1]\n",
    "            try : \n",
    "                rna_type = re.findall(r'([A-Za-z0-9]+\\-)\\)',str(description.split(' ; ')[0]))[0]\n",
    "            except : \n",
    "                rna_type = str(description.split(' ; ')[0])\n",
    "        else : \n",
    "            Start = res[1]\n",
    "            End = res[2]\n",
    "        \n",
    "        rna_annotations.append([rna_id, rna_name, sgd_id, rna_type, Chr, Start, End, description, genome_version])\n",
    "\n",
    "RNA_TABLE = pd.DataFrame(rna_annotations, columns=['locus_id', 'name', 'sgd_id', 'genome_annotations', 'chrom', 'start', 'end', 'description', 'genome_version'])\n",
    "RNA_TABLE['chrom'] = RNA_TABLE['chrom'].apply(format_roman_chrom)\n",
    "RNA_TABLE.to_csv('../../data/genome_annotations/sgd_database/rna_coding_R64-3-1.csv', index=False)\n",
    "RNA_dict = RNA_TABLE.set_index('locus_id').to_dict()\n",
    "\n",
    "#### INTERGENIC\n",
    "\n",
    "intergenic_annotations = []\n",
    "fasta_gz = f'{genome_annot_sgd_folder}/NotFeature_R64-3-1_20210421.fasta.gz'\n",
    "with gzip.open(fasta_gz, \"rt\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        ID, location, genome_version, description = record.description.split(', ')\n",
    "        intergenic_id = ID\n",
    "        res = re.findall(r'(Chr [A-Za-z]+) from ([0-9]+)-([0-9]+)', str(location))[0]\n",
    "        Chr = res[0].upper().replace(' ','_').replace('MITO', 'MT')\n",
    "        Start = res[1]\n",
    "        End = res[2]\n",
    "        \n",
    "        intergenic_annotations.append([intergenic_id, Chr, Start, End, description, genome_version])\n",
    "\n",
    "INTERGENIC_TABLE = pd.DataFrame(intergenic_annotations, columns=['locus_id', 'chrom', 'start', 'end', 'description', 'genome_version'])\n",
    "INTERGENIC_TABLE['chrom'] = INTERGENIC_TABLE['chrom'].apply(format_roman_chrom)\n",
    "INTERGENIC_TABLE.to_csv('../../data/genome_annotations/sgd_database/notFeature_R64-3-1.csv', index=False)\n",
    "INTERGENIC_dict = INTERGENIC_TABLE.set_index('locus_id').to_dict()\n",
    "\n",
    "#### CDS \n",
    "cds_annotations = []\n",
    "fasta_gz = f'{genome_annot_sgd_folder}/orf_coding_all_R64-3-1_20210421.fasta.gz'\n",
    "with gzip.open(fasta_gz, \"rt\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        ID, location, genome_version, cds_type = record.description.split(', ')[:4]\n",
    "        cds_id = ID.split(' ')[0]\n",
    "        cds_name = ID.split(' ')[1]\n",
    "        sgd_id = ID.split(' ')[2]\n",
    "        res = re.findall(r'(Chr [A-Za-z]+) from ([0-9]+)-([0-9]+)', str(location))[0]\n",
    "        Chr = res[0].upper().replace(' ','_').replace('MITO', 'M')\n",
    "        description = ' ; '.join(record.description.split(', ')[4:])\n",
    "        if 'reverse complement' == cds_type:\n",
    "            Start = res[2]\n",
    "            End = res[1]\n",
    "            cds_type = description.split(' ; ')[0]\n",
    "            description = ' ; '.join(description.split(' ; ')[1:])\n",
    "        else : \n",
    "            Start = res[1]\n",
    "            End = res[2]\n",
    "        \n",
    "        cds_annotations.append([cds_id, cds_name, sgd_id, cds_type, Chr, Start, End, description, genome_version])\n",
    "        \n",
    "CDS_TABLE = pd.DataFrame(cds_annotations, columns=['locus_id', 'name', 'sgd_id', 'genome_annotations', 'chrom', 'start', 'end', 'description', 'genome_version'])\n",
    "CDS_TABLE['chrom'] = CDS_TABLE['chrom'].apply(format_roman_chrom)\n",
    "CDS_TABLE.to_csv('../../data/genome_annotations/sgd_database/orf_coding_R64-3-1.csv', index=False)\n",
    "CDS_dict = CDS_TABLE.set_index('locus_id').to_dict()\n",
    "\n",
    "\n",
    "#### 1KB-AWAY\n",
    "\n",
    "gene_annotations = []\n",
    "fasta_gz = f'{genome_annot_sgd_folder}/orf_genomic_1000_all.fasta.gz'\n",
    "with gzip.open(fasta_gz, \"rt\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        ID, location, genome_version, cds_type = record.description.split(', ')[:4]\n",
    "        cds_id = ID.split(' ')[0]\n",
    "        cds_name = ID.split(' ')[1]\n",
    "        sgd_id = ID.split(' ')[2]\n",
    "        res = re.findall(r'(Chr [A-Za-z]+) from ([0-9]+)-([0-9]+)', str(location))[0]\n",
    "        Chr = res[0].upper().replace(' ','_').replace('MITO', 'MT')\n",
    "        description = ' ; '.join(record.description.split(', ')[4:])\n",
    "        if 'reverse complement' == cds_type:\n",
    "            Start = res[2]\n",
    "            End = res[1]\n",
    "            cds_type = description.split(' ; ')[0]\n",
    "            description = ' ; '.join(description.split(' ; ')[1:])\n",
    "        else : \n",
    "            Start = res[1]\n",
    "            End = res[2]\n",
    "        \n",
    "        gene_annotations.append([cds_id, cds_name, sgd_id, cds_type, Chr, Start, End, description, genome_version])\n",
    "        \n",
    "GENE_TABLE = pd.DataFrame(gene_annotations, columns=['locus_id', 'name', 'sgd_id', 'genome_annotations', 'chrom', 'start', 'end', 'description', 'genome_version'])\n",
    "GENE_TABLE['chrom'] = GENE_TABLE['chrom'].apply(format_roman_chrom)\n",
    "GENE_TABLE.to_csv('../../data/genome_annotations/sgd_database/orf_genomic_1000_R64-3-1.csv', index=False)\n",
    "GENE_dict = GENE_TABLE.set_index('locus_id').to_dict()\n",
    "\n",
    "#### PROMOTERS \n",
    "PROMOTER_TABLE = pd.read_csv(f'{genome_annot_rossi_folder}/nfr_ndr_feature_positions.csv').rename(columns={\"Chromosome_Name\":\"chrom\", \"systematic_id\":\"locus_id\", \"common_name\":\"name\", \"NFR/NDR_Start\":\"start\", \"NFR/NDR_End\":\"end\", \"nucleosome_stability\":\"description\"})\n",
    "PROMOTER_TABLE['chrom'] = PROMOTER_TABLE['chrom'].apply(format_chrom)\n",
    "PROMOTER_TABLE['start'] = PROMOTER_TABLE['start'].astype(int)\n",
    "PROMOTER_TABLE['end'] = PROMOTER_TABLE['end'].astype(int)\n",
    "PROMOTER_dict = PROMOTER_TABLE[['locus_id', 'name', 'chrom','strand','start', 'end', 'description']].set_index('locus_id').to_dict()\n",
    "\n",
    "\n",
    "#### UNSTABLE TRANSCRIPTS 2009\n",
    "CUTs_2009_annotation = gffpd.read_gff3('../../data/genome_annotations/xu_2009/Xu_2009_CUTs_V64.gff3')\n",
    "CUTs_2009_annotation = CUTs_2009_annotation.df.rename(columns={'seq_id':'chrom'})\n",
    "CUTs_2009_annotation['chrom'] = [ CUTs_2009_annotation['chrom'][idx].replace(\"chr\", 'CHR_') for idx in CUTs_2009_annotation.index ]\n",
    "CUTs_2009_annotation['locus_id'] = [ CUTs_2009_annotation['attributes'][idx].split(';')[1].replace(\"Name=\", '') for idx in CUTs_2009_annotation.index ]\n",
    "CUTs_2009_annotation['description'] = [ 'CUT' for idx in CUTs_2009_annotation.index ]\n",
    "CUTs_2009_annotation['chrom'] = CUTs_2009_annotation['chrom'].apply(format_roman_chrom)\n",
    "CUT_dict = CUTs_2009_annotation[['locus_id', 'chrom','strand','start', 'end', 'description']].set_index('locus_id').to_dict()\n",
    "\n",
    "\n",
    "SUTs_2009_annotation = gffpd.read_gff3('../../data/genome_annotations/xu_2009/Xu_2009_SUTs_V64.gff3')\n",
    "SUTs_2009_annotation = SUTs_2009_annotation.df.rename(columns={'seq_id':'chrom'})\n",
    "SUTs_2009_annotation['chrom'] = [ SUTs_2009_annotation['chrom'][idx].replace(\"chr\", 'CHR_') for idx in SUTs_2009_annotation.index ]\n",
    "SUTs_2009_annotation['locus_id'] = [ SUTs_2009_annotation['attributes'][idx].split(';')[1].replace(\"Name=\", '') for idx in SUTs_2009_annotation.index ]\n",
    "SUTs_2009_annotation['description'] = [ 'SUT' for idx in SUTs_2009_annotation.index ]\n",
    "SUTs_2009_annotation['chrom'] = SUTs_2009_annotation['chrom'].apply(format_roman_chrom)\n",
    "SUT_dict = SUTs_2009_annotation[['locus_id', 'chrom','strand','start', 'end', 'description']].set_index('locus_id').to_dict()\n",
    "\n",
    "#### UNSTABLE TRANSCRIPTS 2011\n",
    "XUTs_2011_annotation = gffpd.read_gff3('../../data/genome_annotations/van_dikj_2011/van_Dijk_2011_XUTs_V64.gff3')\n",
    "XUTs_2011_annotation = XUTs_2011_annotation.df.rename(columns={'seq_id':'chrom'})\n",
    "XUTs_2011_annotation['chrom'] = [ XUTs_2011_annotation['chrom'][idx].replace(\"chr\", 'CHR_') for idx in XUTs_2011_annotation.index ]\n",
    "XUTs_2011_annotation['locus_id'] = [ XUTs_2011_annotation['attributes'][idx].split(';')[0].replace(\"ID=\", 'XUT_') for idx in XUTs_2011_annotation.index ]\n",
    "XUTs_2011_annotation['description'] = [ XUTs_2011_annotation['attributes'][idx].split(';')[-1].replace(\"desc=\", '') for idx in XUTs_2011_annotation.index ]\n",
    "XUTs_2011_annotation['chrom'] = XUTs_2011_annotation['chrom'].apply(format_roman_chrom)\n",
    "XUT_dict = XUTs_2011_annotation[['locus_id', 'chrom','strand','start', 'end', 'description']].set_index('locus_id').to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_____"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_position(annotation_type, annotation_dict, candidate, snp_pos):\n",
    "    if 'Promoter' in annotation_type : \n",
    "        cdd_strand = annotation_dict['strand'][candidate]\n",
    "        if cdd_strand == '+':\n",
    "            cdd_start = annotation_dict['start'][candidate]\n",
    "            cdd_end = annotation_dict['end'][candidate]\n",
    "        elif cdd_strand == '-':\n",
    "            cdd_start = annotation_dict['end'][candidate]\n",
    "            cdd_end = annotation_dict['start'][candidate]\n",
    "    elif 'Unstable Transcript' in annotation_type : \n",
    "        cdd_strand = annotation_dict['strand'][candidate]\n",
    "        if cdd_strand == '+':\n",
    "            cdd_start = annotation_dict['start'][candidate]\n",
    "            cdd_end = annotation_dict['end'][candidate]\n",
    "        elif cdd_strand == '-':\n",
    "            cdd_start = annotation_dict['end'][candidate]\n",
    "            cdd_end = annotation_dict['start'][candidate]\n",
    "    else :\n",
    "        cdd_start = annotation_dict['start'][candidate]\n",
    "        cdd_end = annotation_dict['end'][candidate]\n",
    "        if '1kb-away' in annotation_type :\n",
    "            distance_start = np.abs(int(snp_pos) - int(cdd_start))\n",
    "            distance_end = np.abs(int(snp_pos) - int(cdd_end))\n",
    "            if min([distance_start,distance_end]) == distance_start:\n",
    "                annotation_type = 'Close to 5\\'-UTR'\n",
    "            if min([distance_start,distance_end]) == distance_end:\n",
    "                annotation_type = 'Close to 3\\'-UTR'\n",
    "\n",
    "    if (int(snp_pos) >= int(cdd_start)) & (int(snp_pos) <= int(cdd_end)):\n",
    "        return True, annotation_type \n",
    "    return False, annotation_type\n",
    "\n",
    "\n",
    "def isAnnotated(snp_id, args):\n",
    "    snps_dict = args[0]\n",
    "    annotation_dict = args[1]\n",
    "    snps_class = args[2]\n",
    "\n",
    "    snp_chrom = snps_dict['chrom'][snp_id]\n",
    "    snp_pos = snps_dict['position'][snp_id]\n",
    "    \n",
    "    candidates = []\n",
    "    for key, val in annotation_dict['chrom'].items():\n",
    "        if snp_chrom == val : \n",
    "            candidates.append(key)\n",
    "    for cdt in candidates :\n",
    "        RES, snps_class = check_position(snps_class, annotation_dict, cdt, snp_pos)\n",
    "        if RES == True :\n",
    "            try: \n",
    "                return (int(snp_id), cdt, annotation_dict[\"name\"][cdt], annotation_dict[\"sgd_id\"][cdt], annotation_dict[\"description\"][cdt], snps_class, annotation_dict[\"genome_annotations\"][cdt])\n",
    "            except: \n",
    "                try :\n",
    "                    return (int(snp_id), cdt, np.nan, np.nan,  annotation_dict[\"description\"][cdt], snps_class, snps_class)\n",
    "                except:\n",
    "                    return (int(snp_id), cdt, np.nan, np.nan,  np.nan, snps_class, snps_class)\n",
    "    return (np.nan, np.nan, np.nan, np.nan, np.nan, np.nan, np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_snps = pd.read_csv('../../data/genotype_information/yeast_snps_loc_dec2022.txt') #### need to be generated by 00a_generate_genotype_matrix.py\n",
    "all_snps['position'] = all_snps['position'].replace(0,1)\n",
    "all_snps['chrom'] = all_snps['chrom'].apply(format_chrom)\n",
    "all_snps_dict = all_snps.set_index('snp_id').to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 12054/12054 [02:23<00:00, 84.21it/s] \n"
     ]
    }
   ],
   "source": [
    "annotations = []\n",
    "for snp_id in tqdm(all_snps_dict['chrom'].keys()):\n",
    "    annotations.append(isAnnotated(snp_id, (all_snps_dict, INTERGENIC_dict, 'Intergenic region')))\n",
    "    annotations.append(isAnnotated(snp_id, (all_snps_dict, GENE_dict, '1kb-away')))\n",
    "    annotations.append(isAnnotated(snp_id, (all_snps_dict, PROMOTER_dict, 'Promoter')))\n",
    "    annotations.append(isAnnotated(snp_id, (all_snps_dict, CDS_dict, 'ORF')))\n",
    "    annotations.append(isAnnotated(snp_id, (all_snps_dict, RNA_dict, 'Non-coding RNA')))\n",
    "    annotations.append(isAnnotated(snp_id, (all_snps_dict, CUT_dict, 'CUT')))\n",
    "    annotations.append(isAnnotated(snp_id, (all_snps_dict, SUT_dict, 'SUT')))\n",
    "    annotations.append(isAnnotated(snp_id, (all_snps_dict, XUT_dict, 'XUT')))\n",
    "    annotations.append(isAnnotated(snp_id, (all_snps_dict, OTHERS_dict, 'Other features')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_ANNOTATIONS = pd.DataFrame(annotations, columns=['snp_id', 'locus_id', 'name', 'sgd_id', 'description','snps_class_up', 'genome_annotations']).dropna(how='all').reset_index(drop=True) # .drop_duplicates(subset=['snp_id'], keep='last')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>name</th>\n",
       "      <th>sgd_id</th>\n",
       "      <th>description</th>\n",
       "      <th>snps_class_up</th>\n",
       "      <th>genome_annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>A:802-1806</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between TEL01L and YAL068C</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>SGDID:S000002143</td>\n",
       "      <td>\"Dubious open reading frame; unlikely to encod...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Dubious ORF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>X1L_NFR/NDR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.0</td>\n",
       "      <td>A:802-1806</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between TEL01L and YAL068C</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.0</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>SGDID:S000002143</td>\n",
       "      <td>\"Dubious open reading frame; unlikely to encod...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Dubious ORF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29929</th>\n",
       "      <td>12050.0</td>\n",
       "      <td>Q0250</td>\n",
       "      <td>COX2</td>\n",
       "      <td>SGDID:S000007281</td>\n",
       "      <td>\"Subunit II of cytochrome c oxidase (Complex I...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Verified ORF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29930</th>\n",
       "      <td>12051.0</td>\n",
       "      <td>Q0250</td>\n",
       "      <td>COX2</td>\n",
       "      <td>SGDID:S000007281</td>\n",
       "      <td>\"Subunit II of cytochrome c oxidase (Complex I...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Verified ORF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29931</th>\n",
       "      <td>12052.0</td>\n",
       "      <td>Q:77506-78088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between YNCQ0023W and YNCQ0024C</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29932</th>\n",
       "      <td>12053.0</td>\n",
       "      <td>Q:82601-85034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between ORI5 and YNCQ0026W</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29933</th>\n",
       "      <td>12054.0</td>\n",
       "      <td>Q:82601-85034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between ORI5 and YNCQ0026W</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29934 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        snp_id       locus_id     name            sgd_id  \\\n",
       "0          1.0     A:802-1806      NaN               NaN   \n",
       "1          1.0        YAL069W  YAL069W  SGDID:S000002143   \n",
       "2          1.0    X1L_NFR/NDR      NaN               NaN   \n",
       "3          2.0     A:802-1806      NaN               NaN   \n",
       "4          2.0        YAL069W  YAL069W  SGDID:S000002143   \n",
       "...        ...            ...      ...               ...   \n",
       "29929  12050.0          Q0250     COX2  SGDID:S000007281   \n",
       "29930  12051.0          Q0250     COX2  SGDID:S000007281   \n",
       "29931  12052.0  Q:77506-78088      NaN               NaN   \n",
       "29932  12053.0  Q:82601-85034      NaN               NaN   \n",
       "29933  12054.0  Q:82601-85034      NaN               NaN   \n",
       "\n",
       "                                             description      snps_class_up  \\\n",
       "0                             between TEL01L and YAL068C  Intergenic region   \n",
       "1      \"Dubious open reading frame; unlikely to encod...    Close to 3'-UTR   \n",
       "2                                                    NaN           Promoter   \n",
       "3                             between TEL01L and YAL068C  Intergenic region   \n",
       "4      \"Dubious open reading frame; unlikely to encod...    Close to 3'-UTR   \n",
       "...                                                  ...                ...   \n",
       "29929  \"Subunit II of cytochrome c oxidase (Complex I...    Close to 3'-UTR   \n",
       "29930  \"Subunit II of cytochrome c oxidase (Complex I...    Close to 3'-UTR   \n",
       "29931                    between YNCQ0023W and YNCQ0024C  Intergenic region   \n",
       "29932                         between ORI5 and YNCQ0026W  Intergenic region   \n",
       "29933                         between ORI5 and YNCQ0026W  Intergenic region   \n",
       "\n",
       "      genome_annotations  \n",
       "0      Intergenic region  \n",
       "1            Dubious ORF  \n",
       "2               Promoter  \n",
       "3      Intergenic region  \n",
       "4            Dubious ORF  \n",
       "...                  ...  \n",
       "29929       Verified ORF  \n",
       "29930       Verified ORF  \n",
       "29931  Intergenic region  \n",
       "29932  Intergenic region  \n",
       "29933  Intergenic region  \n",
       "\n",
       "[29934 rows x 7 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_ANNOTATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_annotations = [] \n",
    "for snp_id in np.unique(ALL_ANNOTATIONS['snp_id']):\n",
    "    TMP = ALL_ANNOTATIONS[ALL_ANNOTATIONS['snp_id'] == snp_id]\n",
    "    snps_annots = []\n",
    "    for snps_class in TMP['snps_class_up'] :\n",
    "        if ('ORF' in snps_class) or ('CUT' in snps_class) or ('SUT' in snps_class) or ('XUT' in snps_class):\n",
    "            snps_annots.append(snps_class)\n",
    "    snps_annots = '_'.join(snps_annots)\n",
    "    snp_annotations.append([int(snp_id), snps_annots])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_stats = pd.DataFrame(snp_annotations, columns=['snp_id', 'snps_annotations']) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['', 'CUT', 'CUT_SUT', 'CUT_SUT_XUT', 'CUT_XUT', 'ORF', 'ORF_CUT',\n",
       "       'ORF_CUT_XUT', 'ORF_SUT', 'ORF_SUT_XUT', 'ORF_XUT', 'SUT',\n",
       "       'SUT_XUT', 'XUT'], dtype=object)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(annotation_stats['snps_annotations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_stats[annotation_stats['snps_annotations'] != ''].to_csv('../../data/genotype_information/snps_annotations_genome-version-3-64-1_UTs.txt', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_name(x):\n",
    "    try : \n",
    "        return GENE_TABLE[GENE_TABLE['locus_id'] == x ]['name'].values[0]\n",
    "    except : \n",
    "        return ALL_ANNOTATIONS[ALL_ANNOTATIONS['locus_id'] == x ]['name'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_ANNOTATIONS['name'] = ALL_ANNOTATIONS['locus_id'].apply(get_gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>name</th>\n",
       "      <th>sgd_id</th>\n",
       "      <th>description</th>\n",
       "      <th>snps_class_up</th>\n",
       "      <th>genome_annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>X1L_NFR/NDR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>X1L_NFR/NDR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>Promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>SGDID:S000002143</td>\n",
       "      <td>\"Dubious open reading frame; unlikely to encod...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Dubious ORF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>SGDID:S000002143</td>\n",
       "      <td>\"Dubious open reading frame; unlikely to encod...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Dubious ORF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>SGDID:S000002143</td>\n",
       "      <td>\"Dubious open reading frame; unlikely to encod...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Dubious ORF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12049</th>\n",
       "      <td>12050.0</td>\n",
       "      <td>Q0250</td>\n",
       "      <td>COX2</td>\n",
       "      <td>SGDID:S000007281</td>\n",
       "      <td>\"Subunit II of cytochrome c oxidase (Complex I...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Verified ORF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12050</th>\n",
       "      <td>12051.0</td>\n",
       "      <td>Q0250</td>\n",
       "      <td>COX2</td>\n",
       "      <td>SGDID:S000007281</td>\n",
       "      <td>\"Subunit II of cytochrome c oxidase (Complex I...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Verified ORF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12051</th>\n",
       "      <td>12052.0</td>\n",
       "      <td>Q:77506-78088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between YNCQ0023W and YNCQ0024C</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12052</th>\n",
       "      <td>12053.0</td>\n",
       "      <td>Q:82601-85034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between ORI5 and YNCQ0026W</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12053</th>\n",
       "      <td>12054.0</td>\n",
       "      <td>Q:82601-85034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between ORI5 and YNCQ0026W</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12054 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        snp_id       locus_id     name            sgd_id  \\\n",
       "0          1.0    X1L_NFR/NDR      NaN               NaN   \n",
       "1          2.0    X1L_NFR/NDR      NaN               NaN   \n",
       "2          3.0        YAL069W  YAL069W  SGDID:S000002143   \n",
       "3          4.0        YAL069W  YAL069W  SGDID:S000002143   \n",
       "4          5.0        YAL069W  YAL069W  SGDID:S000002143   \n",
       "...        ...            ...      ...               ...   \n",
       "12049  12050.0          Q0250     COX2  SGDID:S000007281   \n",
       "12050  12051.0          Q0250     COX2  SGDID:S000007281   \n",
       "12051  12052.0  Q:77506-78088      NaN               NaN   \n",
       "12052  12053.0  Q:82601-85034      NaN               NaN   \n",
       "12053  12054.0  Q:82601-85034      NaN               NaN   \n",
       "\n",
       "                                             description      snps_class_up  \\\n",
       "0                                                    NaN           Promoter   \n",
       "1                                                    NaN           Promoter   \n",
       "2      \"Dubious open reading frame; unlikely to encod...    Close to 3'-UTR   \n",
       "3      \"Dubious open reading frame; unlikely to encod...    Close to 3'-UTR   \n",
       "4      \"Dubious open reading frame; unlikely to encod...    Close to 3'-UTR   \n",
       "...                                                  ...                ...   \n",
       "12049  \"Subunit II of cytochrome c oxidase (Complex I...    Close to 3'-UTR   \n",
       "12050  \"Subunit II of cytochrome c oxidase (Complex I...    Close to 3'-UTR   \n",
       "12051                    between YNCQ0023W and YNCQ0024C  Intergenic region   \n",
       "12052                         between ORI5 and YNCQ0026W  Intergenic region   \n",
       "12053                         between ORI5 and YNCQ0026W  Intergenic region   \n",
       "\n",
       "      genome_annotations  \n",
       "0               Promoter  \n",
       "1               Promoter  \n",
       "2            Dubious ORF  \n",
       "3            Dubious ORF  \n",
       "4            Dubious ORF  \n",
       "...                  ...  \n",
       "12049       Verified ORF  \n",
       "12050       Verified ORF  \n",
       "12051  Intergenic region  \n",
       "12052  Intergenic region  \n",
       "12053  Intergenic region  \n",
       "\n",
       "[12054 rows x 7 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_ANNOTATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>name</th>\n",
       "      <th>sgd_id</th>\n",
       "      <th>description</th>\n",
       "      <th>genome_annotations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snps_class_up</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Close to 3'-UTR</th>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intergenic region</th>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Non-coding RNA</th>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF</th>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other features</th>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter</th>\n",
       "      <td>2276</td>\n",
       "      <td>2276</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1678</td>\n",
       "      <td>2276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unstable transcript</th>\n",
       "      <td>619</td>\n",
       "      <td>619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>619</td>\n",
       "      <td>619</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     snp_id  locus_id  name  sgd_id  description  \\\n",
       "snps_class_up                                                      \n",
       "Close to 3'-UTR        2114      2114  2114    2114         2114   \n",
       "Intergenic region       199       199     0       0          199   \n",
       "Non-coding RNA           55        55    55      55           55   \n",
       "ORF                    6314      6314  6314    6314         6314   \n",
       "Other features          477       477   477     477          477   \n",
       "Promoter               2276      2276     0       0         1678   \n",
       "Unstable transcript     619       619     0       0          619   \n",
       "\n",
       "                     genome_annotations  \n",
       "snps_class_up                            \n",
       "Close to 3'-UTR                    2114  \n",
       "Intergenic region                   199  \n",
       "Non-coding RNA                       55  \n",
       "ORF                                6314  \n",
       "Other features                      477  \n",
       "Promoter                           2276  \n",
       "Unstable transcript                 619  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_ANNOTATIONS.groupby('snps_class_up').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_snps_class_down(x):\n",
    "    locus_id = ALL_ANNOTATIONS[ALL_ANNOTATIONS['snp_id'] == x]['locus_id'].values[0]\n",
    "    snps_class = ALL_ANNOTATIONS[ALL_ANNOTATIONS['snp_id'] == x]['snps_class_up'].values[0]\n",
    "    genome_annotations = ALL_ANNOTATIONS[ALL_ANNOTATIONS['snp_id'] == x]['genome_annotations'].values[0]\n",
    "    if snps_class == 'ORF' :\n",
    "        if '_' in genome_annotations :\n",
    "            genome_annotations = genome_annotations.replace('_',' ')\n",
    "        return genome_annotations\n",
    "    if snps_class == 'Intergenic region':\n",
    "        return 'Intergenic region'\n",
    "    if snps_class == 'Close to 5\\'-UTR':\n",
    "        return 'Close to 5\\'-UTR'\n",
    "    if snps_class == 'Close to 3\\'-UTR':\n",
    "        return 'Close to 3\\'-UTR'\n",
    "    if snps_class == 'Promoter':\n",
    "        return 'Promoter'\n",
    "    if snps_class == 'Non-coding RNA':\n",
    "        if '_' in genome_annotations :\n",
    "            genome_annotations = genome_annotations.replace('_',' ')\n",
    "        return genome_annotations\n",
    "    if 'UT' in locus_id : \n",
    "        try : \n",
    "            x = re.findall(r'^([CSXUT]{3})', locus_id)[0]\n",
    "            return x \n",
    "        except: \n",
    "            return locus_id \n",
    "    if 'ARS' in locus_id : \n",
    "        return 'Replication' \n",
    "    if 'CEN' in locus_id : \n",
    "        return \"Centromere\"\n",
    "    if 'TEL' in locus_id : \n",
    "        return \"Telomere\"\n",
    "    if 'NTS1-2' in locus_id :\n",
    "        return 'Mating-related region'\n",
    "    if 'RE' in locus_id :\n",
    "        return 'Recombination enhancer'\n",
    "    if 'HM' in locus_id :\n",
    "        return \"Mating-related region\"\n",
    "    if 'MATALPHA' in locus_id :\n",
    "        return \"Mating-related region\"\n",
    "    else : \n",
    "        try : \n",
    "            res = re.findall(r'(^[A-Z]+[a-z]+)', locus_id)\n",
    "            if len(res) == 1 :\n",
    "                if genome_annotations == \"LTR_retrotransposon\":\n",
    "                    return \"LTR retrotransposon\"\n",
    "                else : \n",
    "                    return \"LTR\"\n",
    "        except :\n",
    "            return x \n",
    "    return np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 12054/12054 [00:34<00:00, 351.90it/s]\n"
     ]
    }
   ],
   "source": [
    "ALL_ANNOTATIONS[\"snps_class_down\"] = ALL_ANNOTATIONS['snp_id'].progress_apply(get_snps_class_down)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>name</th>\n",
       "      <th>sgd_id</th>\n",
       "      <th>description</th>\n",
       "      <th>genome_annotations</th>\n",
       "      <th>snps_class_down</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snps_class_up</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Close to 3'-UTR</th>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intergenic region</th>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Non-coding RNA</th>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ORF</th>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "      <td>6314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other features</th>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter</th>\n",
       "      <td>2276</td>\n",
       "      <td>2276</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1678</td>\n",
       "      <td>2276</td>\n",
       "      <td>2276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unstable transcript</th>\n",
       "      <td>619</td>\n",
       "      <td>619</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>619</td>\n",
       "      <td>619</td>\n",
       "      <td>619</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     snp_id  locus_id  name  sgd_id  description  \\\n",
       "snps_class_up                                                      \n",
       "Close to 3'-UTR        2114      2114  2114    2114         2114   \n",
       "Intergenic region       199       199     0       0          199   \n",
       "Non-coding RNA           55        55    55      55           55   \n",
       "ORF                    6314      6314  6314    6314         6314   \n",
       "Other features          477       477   477     477          477   \n",
       "Promoter               2276      2276     0       0         1678   \n",
       "Unstable transcript     619       619     0       0          619   \n",
       "\n",
       "                     genome_annotations  snps_class_down  \n",
       "snps_class_up                                             \n",
       "Close to 3'-UTR                    2114             2114  \n",
       "Intergenic region                   199              199  \n",
       "Non-coding RNA                       55               55  \n",
       "ORF                                6314             6314  \n",
       "Other features                      477              477  \n",
       "Promoter                           2276             2276  \n",
       "Unstable transcript                 619              619  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_ANNOTATIONS.groupby(\"snps_class_up\").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>name</th>\n",
       "      <th>sgd_id</th>\n",
       "      <th>description</th>\n",
       "      <th>snps_class_up</th>\n",
       "      <th>genome_annotations</th>\n",
       "      <th>snps_class_down</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [snp_id, locus_id, name, sgd_id, description, snps_class_up, genome_annotations, snps_class_down]\n",
       "Index: []"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_ANNOTATIONS[(ALL_ANNOTATIONS['snps_class_up'] == 'CDS') & (ALL_ANNOTATIONS['snps_class_down'] != 'CDS')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>name</th>\n",
       "      <th>sgd_id</th>\n",
       "      <th>description</th>\n",
       "      <th>snps_class_up</th>\n",
       "      <th>genome_annotations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snps_class_down</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CUT</th>\n",
       "      <td>109</td>\n",
       "      <td>109</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>109</td>\n",
       "      <td>109</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Centromere</th>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Close to 3'-UTR</th>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "      <td>2114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dubious ORF</th>\n",
       "      <td>202</td>\n",
       "      <td>202</td>\n",
       "      <td>202</td>\n",
       "      <td>202</td>\n",
       "      <td>202</td>\n",
       "      <td>202</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intergenic region</th>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LTR</th>\n",
       "      <td>161</td>\n",
       "      <td>161</td>\n",
       "      <td>161</td>\n",
       "      <td>161</td>\n",
       "      <td>161</td>\n",
       "      <td>161</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LTR retrotransposon</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mating-related region</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter</th>\n",
       "      <td>2276</td>\n",
       "      <td>2276</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1678</td>\n",
       "      <td>2276</td>\n",
       "      <td>2276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Recombination enhancer</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Replication</th>\n",
       "      <td>204</td>\n",
       "      <td>204</td>\n",
       "      <td>204</td>\n",
       "      <td>204</td>\n",
       "      <td>204</td>\n",
       "      <td>204</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUT</th>\n",
       "      <td>332</td>\n",
       "      <td>332</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>332</td>\n",
       "      <td>332</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Telomere</th>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uncharacterized ORF</th>\n",
       "      <td>582</td>\n",
       "      <td>582</td>\n",
       "      <td>582</td>\n",
       "      <td>582</td>\n",
       "      <td>582</td>\n",
       "      <td>582</td>\n",
       "      <td>582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Verified ORF</th>\n",
       "      <td>5514</td>\n",
       "      <td>5514</td>\n",
       "      <td>5514</td>\n",
       "      <td>5514</td>\n",
       "      <td>5514</td>\n",
       "      <td>5514</td>\n",
       "      <td>5514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XUT</th>\n",
       "      <td>178</td>\n",
       "      <td>178</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>178</td>\n",
       "      <td>178</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>blocked reading frame</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA gene</th>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogene</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snoRNA gene</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tRNA gene</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>telomerase RNA gene</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        snp_id  locus_id  name  sgd_id  description  \\\n",
       "snps_class_down                                                       \n",
       "CUT                        109       109     0       0          109   \n",
       "Centromere                  17        17    17      17           17   \n",
       "Close to 3'-UTR           2114      2114  2114    2114         2114   \n",
       "Dubious ORF                202       202   202     202          202   \n",
       "Intergenic region          199       199     0       0          199   \n",
       "LTR                        161       161   161     161          161   \n",
       "LTR retrotransposon         14        14    14      14           14   \n",
       "Mating-related region       12        12    12      12           12   \n",
       "Promoter                  2276      2276     0       0         1678   \n",
       "Recombination enhancer       1         1     1       1            1   \n",
       "Replication                204       204   204     204          204   \n",
       "SUT                        332       332     0       0          332   \n",
       "Telomere                    68        68    68      68           68   \n",
       "Uncharacterized ORF        582       582   582     582          582   \n",
       "Verified ORF              5514      5514  5514    5514         5514   \n",
       "XUT                        178       178     0       0          178   \n",
       "blocked reading frame        4         4     4       4            4   \n",
       "ncRNA gene                  39        39    39      39           39   \n",
       "pseudogene                  12        12    12      12           12   \n",
       "snoRNA gene                 11        11    11      11           11   \n",
       "tRNA gene                    4         4     4       4            4   \n",
       "telomerase RNA gene          1         1     1       1            1   \n",
       "\n",
       "                        snps_class_up  genome_annotations  \n",
       "snps_class_down                                            \n",
       "CUT                               109                 109  \n",
       "Centromere                         17                  17  \n",
       "Close to 3'-UTR                  2114                2114  \n",
       "Dubious ORF                       202                 202  \n",
       "Intergenic region                 199                 199  \n",
       "LTR                               161                 161  \n",
       "LTR retrotransposon                14                  14  \n",
       "Mating-related region              12                  12  \n",
       "Promoter                         2276                2276  \n",
       "Recombination enhancer              1                   1  \n",
       "Replication                       204                 204  \n",
       "SUT                               332                 332  \n",
       "Telomere                           68                  68  \n",
       "Uncharacterized ORF               582                 582  \n",
       "Verified ORF                     5514                5514  \n",
       "XUT                               178                 178  \n",
       "blocked reading frame               4                   4  \n",
       "ncRNA gene                         39                  39  \n",
       "pseudogene                         12                  12  \n",
       "snoRNA gene                        11                  11  \n",
       "tRNA gene                           4                   4  \n",
       "telomerase RNA gene                 1                   1  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALL_ANNOTATIONS.groupby(\"snps_class_down\").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_snps.merge(ALL_ANNOTATIONS, on='snp_id', how='outer').to_csv('../../data/genotype_information/snps_annotations_genome-version-3-64-1.txt', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('piQTL')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13 (default, Mar 28 2022, 11:38:47) \n[GCC 7.5.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b1c9f147d5e11ed2bfef82e704687f362a942886afe9a4c3e913143ced90e071"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
