{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cv2, os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from assocplots.manhattan import *\n",
    "from assocplots.qqplot import *\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### SNPs annotations\n",
    "snps_loc = pd.read_csv('../../data/genotype_information/snps_annotations_genome-version-3-64-1.txt').rename(columns={'snp_id':'SNP'})\n",
    "snps_loc['SNP'] = snps_loc['SNP'].astype(int)\n",
    "snps_loc['SNP_ID'] = [ f'{snps_loc[\"chrom\"][idx].replace(\"_\",\"\")}_{snps_loc[\"SNP\"][idx]}' for idx in snps_loc.index]\n",
    "snps_loc['position'] = snps_loc['position'].astype(int)\n",
    "\n",
    "#### Gene names \n",
    "gene_loc = pd.read_csv('../../data/genotype_information/yeast_ORFs_dec2022.txt')\n",
    "\n",
    "#### Genotype matrix\n",
    "genotype = pd.read_csv('../../data/genotype_information/piQTL_genotype_matrix_dec2022.txt')\n",
    "nb_strains = len(genotype.columns) - 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>17</th>\n",
       "      <th>20</th>\n",
       "      <th>22</th>\n",
       "      <th>27</th>\n",
       "      <th>29</th>\n",
       "      <th>...</th>\n",
       "      <th>1038</th>\n",
       "      <th>1040</th>\n",
       "      <th>1046</th>\n",
       "      <th>1050</th>\n",
       "      <th>1052</th>\n",
       "      <th>1065</th>\n",
       "      <th>1069</th>\n",
       "      <th>17_2</th>\n",
       "      <th>40_2</th>\n",
       "      <th>180_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12049</th>\n",
       "      <td>12050</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12050</th>\n",
       "      <td>12051</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12051</th>\n",
       "      <td>12052</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12052</th>\n",
       "      <td>12053</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12053</th>\n",
       "      <td>12054</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12054 rows × 358 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       snp_id  6  7  14  15  17  20  22  27  29  ...  1038  1040  1046  1050  \\\n",
       "0           1  1  0   1   0   1   0   1  -1   0  ...    -1    -1    -1    -1   \n",
       "1           2  1  1  -1  -1   1   0   1  -1  -1  ...    -1     0     0    -1   \n",
       "2           3  0  0  -1  -1   0  -1   0   0   0  ...    -1     0     0    -1   \n",
       "3           4  0  0   0   0   0   0   0   0   0  ...     0     0    -1     0   \n",
       "4           5 -1 -1  -1  -1  -1   1  -1  -1  -1  ...     1     0    -1     0   \n",
       "...       ... .. ..  ..  ..  ..  ..  ..  ..  ..  ...   ...   ...   ...   ...   \n",
       "12049   12050  1  1   1  -1  -1   1  -1   1  -1  ...    -1    -1    -1    -1   \n",
       "12050   12051  0  0   0   0   0   0   0   0   0  ...    -1    -1    -1    -1   \n",
       "12051   12052  0  0   1   0   0   0   0   0   0  ...     0     0     0     0   \n",
       "12052   12053  1  0   1   0   0  -1  -1   0  -1  ...     0     0     0     0   \n",
       "12053   12054  1  0   1   0   0  -1  -1   0  -1  ...     0     0     0     0   \n",
       "\n",
       "       1052  1065  1069  17_2  40_2  180_2  \n",
       "0        -1    -1     0     1     1     -1  \n",
       "1         0     0     0     1     1     -1  \n",
       "2         0     0     0     0     0      0  \n",
       "3         0    -1    -1     0     0      0  \n",
       "4         0     1    -1    -1     1     -1  \n",
       "...     ...   ...   ...   ...   ...    ...  \n",
       "12049     1     1     0    -1     1     -1  \n",
       "12050     1    -1     1     0     0     -1  \n",
       "12051     0     0     0     0     0      0  \n",
       "12052     0     1     0     0     0      0  \n",
       "12053     0     1     0     0     0      0  \n",
       "\n",
       "[12054 rows x 358 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_id</th>\n",
       "      <th>name</th>\n",
       "      <th>sgd_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YAL069W</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>SGDID:S000002143</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL068W-A</td>\n",
       "      <td>YAL068W-A</td>\n",
       "      <td>SGDID:S000028594</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>538</td>\n",
       "      <td>792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YAL068C</td>\n",
       "      <td>PAU8</td>\n",
       "      <td>SGDID:S000002142</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>SGDID:S000028593</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YAL067C</td>\n",
       "      <td>SEO1</td>\n",
       "      <td>SGDID:S000000062</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6602</th>\n",
       "      <td>Q0182</td>\n",
       "      <td>Q0182</td>\n",
       "      <td>SGDID:S000007280</td>\n",
       "      <td>CHR_1000</td>\n",
       "      <td>65770</td>\n",
       "      <td>66174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6603</th>\n",
       "      <td>Q0250</td>\n",
       "      <td>COX2</td>\n",
       "      <td>SGDID:S000007281</td>\n",
       "      <td>CHR_1000</td>\n",
       "      <td>73758</td>\n",
       "      <td>74513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6604</th>\n",
       "      <td>Q0255</td>\n",
       "      <td>Q0255</td>\n",
       "      <td>SGDID:S000007282</td>\n",
       "      <td>CHR_1000</td>\n",
       "      <td>74495</td>\n",
       "      <td>75622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6605</th>\n",
       "      <td>Q0275</td>\n",
       "      <td>COX3</td>\n",
       "      <td>SGDID:S000007283</td>\n",
       "      <td>CHR_1000</td>\n",
       "      <td>79213</td>\n",
       "      <td>80022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6606</th>\n",
       "      <td>Q0297</td>\n",
       "      <td>Q0297</td>\n",
       "      <td>SGDID:S000007284</td>\n",
       "      <td>CHR_1000</td>\n",
       "      <td>85554</td>\n",
       "      <td>85709</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6607 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       locus_id       name            sgd_id     chrom  start    end\n",
       "0       YAL069W    YAL069W  SGDID:S000002143     CHR_1    335    649\n",
       "1     YAL068W-A  YAL068W-A  SGDID:S000028594     CHR_1    538    792\n",
       "2       YAL068C       PAU8  SGDID:S000002142     CHR_1   1807   2169\n",
       "3     YAL067W-A  YAL067W-A  SGDID:S000028593     CHR_1   2480   2707\n",
       "4       YAL067C       SEO1  SGDID:S000000062     CHR_1   7235   9016\n",
       "...         ...        ...               ...       ...    ...    ...\n",
       "6602      Q0182      Q0182  SGDID:S000007280  CHR_1000  65770  66174\n",
       "6603      Q0250       COX2  SGDID:S000007281  CHR_1000  73758  74513\n",
       "6604      Q0255      Q0255  SGDID:S000007282  CHR_1000  74495  75622\n",
       "6605      Q0275       COX3  SGDID:S000007283  CHR_1000  79213  80022\n",
       "6606      Q0297      Q0297  SGDID:S000007284  CHR_1000  85554  85709\n",
       "\n",
       "[6607 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_loc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_chr(x, gene_loc):\n",
    "    try : \n",
    "        return gene_loc[gene_loc['name'] == x]['chrom'].values[0].replace('CHR_', '')\n",
    "    except : \n",
    "        return np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "PPI_list = pd.read_csv('../../data/pipeline/PPI_reference_barcodes.csv')\n",
    "PPI_list['PPI_bait'] =  [ PPI_list['PPI'][idx].split('_')[0] for idx in PPI_list.index ]\n",
    "PPI_list['PPI_prey'] =  [ PPI_list['PPI'][idx].split('_')[1] for idx in PPI_list.index ]\n",
    "\n",
    "PPI_list['CHR_bait'] = PPI_list['PPI_bait'].apply(get_chr, args=(gene_loc,))\n",
    "PPI_list['CHR_prey'] = PPI_list['PPI_prey'].apply(get_chr, args=(gene_loc,))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>well_position</th>\n",
       "      <th>ppi</th>\n",
       "      <th>N-index</th>\n",
       "      <th>S-index</th>\n",
       "      <th>N_barcode</th>\n",
       "      <th>S_barcode</th>\n",
       "      <th>PPI</th>\n",
       "      <th>PPI_bait</th>\n",
       "      <th>PPI_prey</th>\n",
       "      <th>CHR_bait</th>\n",
       "      <th>CHR_prey</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1</td>\n",
       "      <td>GTR1:SLM4</td>\n",
       "      <td>N701</td>\n",
       "      <td>S501</td>\n",
       "      <td>TCGCCTTA</td>\n",
       "      <td>TAGATCGC</td>\n",
       "      <td>GTR1_SLM4</td>\n",
       "      <td>GTR1</td>\n",
       "      <td>SLM4</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2</td>\n",
       "      <td>SNA4:GTR2</td>\n",
       "      <td>N702</td>\n",
       "      <td>S501</td>\n",
       "      <td>CTAGTACG</td>\n",
       "      <td>TAGATCGC</td>\n",
       "      <td>SNA4_GTR2</td>\n",
       "      <td>SNA4</td>\n",
       "      <td>GTR2</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A3</td>\n",
       "      <td>ERG11:HYR1</td>\n",
       "      <td>N703</td>\n",
       "      <td>S501</td>\n",
       "      <td>TTCTGCCT</td>\n",
       "      <td>TAGATCGC</td>\n",
       "      <td>ERG11_HYR1</td>\n",
       "      <td>ERG11</td>\n",
       "      <td>HYR1</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A4</td>\n",
       "      <td>DHH1:LSM3</td>\n",
       "      <td>N704</td>\n",
       "      <td>S501</td>\n",
       "      <td>GCTCAGGA</td>\n",
       "      <td>TAGATCGC</td>\n",
       "      <td>DHH1_LSM3</td>\n",
       "      <td>DHH1</td>\n",
       "      <td>LSM3</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A5</td>\n",
       "      <td>MID2:PIN2</td>\n",
       "      <td>N705</td>\n",
       "      <td>S501</td>\n",
       "      <td>AGGAGTCC</td>\n",
       "      <td>TAGATCGC</td>\n",
       "      <td>MID2_PIN2</td>\n",
       "      <td>MID2</td>\n",
       "      <td>PIN2</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>H4</td>\n",
       "      <td>MID2:MAM33</td>\n",
       "      <td>N704</td>\n",
       "      <td>S508</td>\n",
       "      <td>GCTCAGGA</td>\n",
       "      <td>CTAAGCCT</td>\n",
       "      <td>MID2_MAM33</td>\n",
       "      <td>MID2</td>\n",
       "      <td>MAM33</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>H5</td>\n",
       "      <td>MID2:WSC2</td>\n",
       "      <td>N705</td>\n",
       "      <td>S508</td>\n",
       "      <td>AGGAGTCC</td>\n",
       "      <td>CTAAGCCT</td>\n",
       "      <td>MID2_WSC2</td>\n",
       "      <td>MID2</td>\n",
       "      <td>WSC2</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>H7</td>\n",
       "      <td>HNM1:TPO2</td>\n",
       "      <td>N707</td>\n",
       "      <td>S508</td>\n",
       "      <td>GTAGAGAG</td>\n",
       "      <td>CTAAGCCT</td>\n",
       "      <td>HNM1_TPO2</td>\n",
       "      <td>HNM1</td>\n",
       "      <td>TPO2</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>H8</td>\n",
       "      <td>ALO1:ADE17</td>\n",
       "      <td>N708</td>\n",
       "      <td>S508</td>\n",
       "      <td>CCTCTCTG</td>\n",
       "      <td>CTAAGCCT</td>\n",
       "      <td>ALO1_ADE17</td>\n",
       "      <td>ALO1</td>\n",
       "      <td>ADE17</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>G9</td>\n",
       "      <td>noPPI:reference</td>\n",
       "      <td>N709</td>\n",
       "      <td>S507</td>\n",
       "      <td>AGCGTAGC</td>\n",
       "      <td>AAGGAGTA</td>\n",
       "      <td>noPPI_reference</td>\n",
       "      <td>noPPI</td>\n",
       "      <td>reference</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   well_position              ppi N-index S-index N_barcode S_barcode  \\\n",
       "0             A1        GTR1:SLM4    N701    S501  TCGCCTTA  TAGATCGC   \n",
       "1             A2        SNA4:GTR2    N702    S501  CTAGTACG  TAGATCGC   \n",
       "2             A3       ERG11:HYR1    N703    S501  TTCTGCCT  TAGATCGC   \n",
       "3             A4        DHH1:LSM3    N704    S501  GCTCAGGA  TAGATCGC   \n",
       "4             A5        MID2:PIN2    N705    S501  AGGAGTCC  TAGATCGC   \n",
       "..           ...              ...     ...     ...       ...       ...   \n",
       "57            H4       MID2:MAM33    N704    S508  GCTCAGGA  CTAAGCCT   \n",
       "58            H5        MID2:WSC2    N705    S508  AGGAGTCC  CTAAGCCT   \n",
       "59            H7        HNM1:TPO2    N707    S508  GTAGAGAG  CTAAGCCT   \n",
       "60            H8       ALO1:ADE17    N708    S508  CCTCTCTG  CTAAGCCT   \n",
       "61            G9  noPPI:reference    N709    S507  AGCGTAGC  AAGGAGTA   \n",
       "\n",
       "                PPI PPI_bait   PPI_prey CHR_bait CHR_prey  \n",
       "0         GTR1_SLM4     GTR1       SLM4       13        2  \n",
       "1         SNA4_GTR2     SNA4       GTR2        4        7  \n",
       "2        ERG11_HYR1    ERG11       HYR1        8        9  \n",
       "3         DHH1_LSM3     DHH1       LSM3        4       12  \n",
       "4         MID2_PIN2     MID2       PIN2       12       15  \n",
       "..              ...      ...        ...      ...      ...  \n",
       "57       MID2_MAM33     MID2      MAM33       12        9  \n",
       "58        MID2_WSC2     MID2       WSC2       12       14  \n",
       "59        HNM1_TPO2     HNM1       TPO2        7        7  \n",
       "60       ALO1_ADE17     ALO1      ADE17       13       13  \n",
       "61  noPPI_reference    noPPI  reference      NaN      NaN  \n",
       "\n",
       "[62 rows x 11 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PPI_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>locus_id</th>\n",
       "      <th>name</th>\n",
       "      <th>sgd_id</th>\n",
       "      <th>description</th>\n",
       "      <th>snps_class_up</th>\n",
       "      <th>genome_annotations</th>\n",
       "      <th>snps_class_down</th>\n",
       "      <th>SNP_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>1025</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>X1L_NFR/NDR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>CHR1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>1035</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>X1L_NFR/NDR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>Promoter</td>\n",
       "      <td>CHR1_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>1074</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>SGDID:S000002143</td>\n",
       "      <td>\"Dubious open reading frame; unlikely to encod...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Dubious ORF</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>CHR1_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>1470</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>SGDID:S000002143</td>\n",
       "      <td>\"Dubious open reading frame; unlikely to encod...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Dubious ORF</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>CHR1_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CHR_1</td>\n",
       "      <td>1475</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>SGDID:S000002143</td>\n",
       "      <td>\"Dubious open reading frame; unlikely to encod...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Dubious ORF</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>CHR1_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12049</th>\n",
       "      <td>12050</td>\n",
       "      <td>CHR_MT</td>\n",
       "      <td>74432</td>\n",
       "      <td>TG</td>\n",
       "      <td>T</td>\n",
       "      <td>Q0250</td>\n",
       "      <td>COX2</td>\n",
       "      <td>SGDID:S000007281</td>\n",
       "      <td>\"Subunit II of cytochrome c oxidase (Complex I...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Verified ORF</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>CHRMT_12050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12050</th>\n",
       "      <td>12051</td>\n",
       "      <td>CHR_MT</td>\n",
       "      <td>74433</td>\n",
       "      <td>G</td>\n",
       "      <td>GGACTA</td>\n",
       "      <td>Q0250</td>\n",
       "      <td>COX2</td>\n",
       "      <td>SGDID:S000007281</td>\n",
       "      <td>\"Subunit II of cytochrome c oxidase (Complex I...</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>Verified ORF</td>\n",
       "      <td>Close to 3'-UTR</td>\n",
       "      <td>CHRMT_12051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12051</th>\n",
       "      <td>12052</td>\n",
       "      <td>CHR_MT</td>\n",
       "      <td>77708</td>\n",
       "      <td>T</td>\n",
       "      <td>A,G</td>\n",
       "      <td>Q:77506-78088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between YNCQ0023W and YNCQ0024C</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>CHRMT_12052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12052</th>\n",
       "      <td>12053</td>\n",
       "      <td>CHR_MT</td>\n",
       "      <td>84245</td>\n",
       "      <td>G</td>\n",
       "      <td>GCC,GTC</td>\n",
       "      <td>Q:82601-85034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between ORI5 and YNCQ0026W</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>CHRMT_12053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12053</th>\n",
       "      <td>12054</td>\n",
       "      <td>CHR_MT</td>\n",
       "      <td>84250</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Q:82601-85034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>between ORI5 and YNCQ0026W</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>Intergenic region</td>\n",
       "      <td>CHRMT_12054</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12054 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SNP   chrom  position REF      ALT       locus_id     name  \\\n",
       "0          1   CHR_1      1025   C        T    X1L_NFR/NDR      NaN   \n",
       "1          2   CHR_1      1035   A        T    X1L_NFR/NDR      NaN   \n",
       "2          3   CHR_1      1074   C        A        YAL069W  YAL069W   \n",
       "3          4   CHR_1      1470   A        G        YAL069W  YAL069W   \n",
       "4          5   CHR_1      1475  AT        A        YAL069W  YAL069W   \n",
       "...      ...     ...       ...  ..      ...            ...      ...   \n",
       "12049  12050  CHR_MT     74432  TG        T          Q0250     COX2   \n",
       "12050  12051  CHR_MT     74433   G   GGACTA          Q0250     COX2   \n",
       "12051  12052  CHR_MT     77708   T      A,G  Q:77506-78088      NaN   \n",
       "12052  12053  CHR_MT     84245   G  GCC,GTC  Q:82601-85034      NaN   \n",
       "12053  12054  CHR_MT     84250   C        T  Q:82601-85034      NaN   \n",
       "\n",
       "                 sgd_id                                        description  \\\n",
       "0                   NaN                                                NaN   \n",
       "1                   NaN                                                NaN   \n",
       "2      SGDID:S000002143  \"Dubious open reading frame; unlikely to encod...   \n",
       "3      SGDID:S000002143  \"Dubious open reading frame; unlikely to encod...   \n",
       "4      SGDID:S000002143  \"Dubious open reading frame; unlikely to encod...   \n",
       "...                 ...                                                ...   \n",
       "12049  SGDID:S000007281  \"Subunit II of cytochrome c oxidase (Complex I...   \n",
       "12050  SGDID:S000007281  \"Subunit II of cytochrome c oxidase (Complex I...   \n",
       "12051               NaN                    between YNCQ0023W and YNCQ0024C   \n",
       "12052               NaN                         between ORI5 and YNCQ0026W   \n",
       "12053               NaN                         between ORI5 and YNCQ0026W   \n",
       "\n",
       "           snps_class_up genome_annotations    snps_class_down       SNP_ID  \n",
       "0               Promoter           Promoter           Promoter       CHR1_1  \n",
       "1               Promoter           Promoter           Promoter       CHR1_2  \n",
       "2        Close to 3'-UTR        Dubious ORF    Close to 3'-UTR       CHR1_3  \n",
       "3        Close to 3'-UTR        Dubious ORF    Close to 3'-UTR       CHR1_4  \n",
       "4        Close to 3'-UTR        Dubious ORF    Close to 3'-UTR       CHR1_5  \n",
       "...                  ...                ...                ...          ...  \n",
       "12049    Close to 3'-UTR       Verified ORF    Close to 3'-UTR  CHRMT_12050  \n",
       "12050    Close to 3'-UTR       Verified ORF    Close to 3'-UTR  CHRMT_12051  \n",
       "12051  Intergenic region  Intergenic region  Intergenic region  CHRMT_12052  \n",
       "12052  Intergenic region  Intergenic region  Intergenic region  CHRMT_12053  \n",
       "12053  Intergenic region  Intergenic region  Intergenic region  CHRMT_12054  \n",
       "\n",
       "[12054 rows x 13 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_loc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 620/620 [03:47<00:00,  2.72it/s]\n"
     ]
    }
   ],
   "source": [
    "input_path = f'../../results/04a_rMVP/rMVP_outputs/piQTL_dec_2022/'\n",
    "output_path = f'../../results/04a_rMVP/rMVP_outputs/piQTL_dec_2022_annotated/'\n",
    "for table in tqdm(os.listdir(os.path.join(input_path))) :\n",
    "    qtl_res = pd.read_csv(os.path.join(input_path, table))\n",
    "    qtl_res = qtl_res.merge(snps_loc[['SNP', 'locus_id', 'name','sgd_id', 'snps_class_up', 'snps_class_down', 'description',  'genome_annotations']], on='SNP')\n",
    "    qtl_res.to_csv(os.path.join(output_path, table), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cp ../../results/04a_rMVP/rMVP_outputs/piQTL_dec_2022_annotated/*.csv /home/savvy/PROJECTS/PHD/piQTL/data/QTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cp ../../results/04a_rMVP/rMVP_outputs/piQTL_dec_2022_annotated/*.csv /home/savvy/PROJECTS/PHD/piQTL/data/QTL_v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_path = '/home/savvy/PROJECTS/PHD/piQTL/data/QTL'\n",
    "for PPI in tqdm(PPI_list['PPI']): \n",
    "    for DRUG in ['noDrug', '5.FC', 'Fluconazole', 'Metformin', 'Trifluoperazine'] : \n",
    "        # print(PPI)\n",
    "        CHR_bait =  PPI_list[PPI_list['PPI'] == PPI]['CHR_bait'].values[0] \n",
    "        CHR_prey = PPI_list[PPI_list['PPI'] == PPI]['CHR_prey'].values[0] \n",
    "        # print(CHR_prey, CHR_bait)\n",
    "\n",
    "        MTX = pd.read_csv(os.path.join(folder_path,f'{PPI}_MTX_{DRUG}_avg_logratio_Fitness_minus_ref.csv')).replace('MT', 17)\n",
    "        MTX = noMTX.rename(columns={'Chr':'CHR', 'Pos':'BP', f'{PPI}_MTX_{DRUG}_avg_logratio_Fitness_minus_ref.GLM':'P', 'Effect':'EFFECTSIZE', 'name':'GENE'})\n",
    "        # MTX.to_csv(os.path.join(folder_path,f'{PPI}_MTX_{DRUG}_avg_logratio_Fitness_minus_ref.csv'), index=False)\n",
    "        \n",
    "        noMTX = pd.read_csv(os.path.join(folder_path,f'{PPI}_noMTX_{DRUG}_avg_logratio_Fitness_minus_ref.csv')).replace('MT', 17)\n",
    "        noMTX = noMTX.rename(columns={'Chr':'CHR', 'Pos':'BP', f'{PPI}_noMTX_{DRUG}_avg_logratio_Fitness_minus_ref.GLM':'P', 'Effect':'EFFECTSIZE', 'name':'GENE'})\n",
    "        # noMTX.to_csv(os.path.join(folder_path,f'{PPI}_noMTX_{DRUG}_avg_logratio_Fitness_minus_ref.csv'), index=False)\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "eQTL_res = pd.read_csv('../../results/04b_eQTL_matrix/piQTL_dec2022_pval_results.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 62/62 [2:39:26<00:00, 154.30s/it]  \n"
     ]
    }
   ],
   "source": [
    "folder_path = '/home/savvy/PROJECTS/PHD/piQTL/data/QTL_v2'\n",
    "for PPI in tqdm(PPI_list['PPI']): \n",
    "    for DRUG in ['noDrug', '5.FC', 'Fluconazole', 'Metformin', 'Trifluoperazine'] : \n",
    "        # print(PPI)\n",
    "        CHR_bait =  PPI_list[PPI_list['PPI'] == PPI]['CHR_bait'].values[0] \n",
    "        CHR_prey = PPI_list[PPI_list['PPI'] == PPI]['CHR_prey'].values[0] \n",
    "        # print(CHR_prey, CHR_bait)\n",
    "\n",
    "        MTX = pd.read_csv(os.path.join(folder_path,f'{PPI}_MTX_{DRUG}_avg_logratio_Fitness_minus_ref.csv')).replace('MT', 17)\n",
    "        eQTL_MTX_RES = eQTL_res[(eQTL_res['gene'].str.contains(PPI)) & (eQTL_res['gene'].str.contains(DRUG)) & (eQTL_res['gene'].str.contains('_MTX'))]\n",
    "        MTX = eQTL_MTX_RES.merge(MTX, on='SNP')\n",
    "        MTX = MTX.rename(columns={'Chr':'CHR', 'Pos':'BP', 'p-value':'P', 'beta':'EFFECTSIZE', 'name':'GENE'})\n",
    "        MTX.to_csv(os.path.join(folder_path,f'{PPI}_MTX_{DRUG}_avg_logratio_Fitness_minus_ref.csv'), index=False)\n",
    "        \n",
    "        noMTX = pd.read_csv(os.path.join(folder_path,f'{PPI}_noMTX_{DRUG}_avg_logratio_Fitness_minus_ref.csv')).replace('MT', 17)\n",
    "        eQTL_noMTX_RES = eQTL_res[(eQTL_res['gene'].str.contains(PPI)) & (eQTL_res['gene'].str.contains(DRUG)) & (eQTL_res['gene'].str.contains('_noMTX'))]\n",
    "        noMTX = eQTL_noMTX_RES.merge(noMTX, on='SNP')\n",
    "        noMTX = MTX.rename(columns={'Chr':'CHR', 'Pos':'BP', 'p-value':'P', 'beta':'EFFECTSIZE', 'name':'GENE'})  \n",
    "        noMTX.to_csv(os.path.join(folder_path,f'{PPI}_noMTX_{DRUG}_avg_logratio_Fitness_minus_ref.csv'), index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('piQTL')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13 (default, Mar 28 2022, 11:38:47) \n[GCC 7.5.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b1c9f147d5e11ed2bfef82e704687f362a942886afe9a4c3e913143ced90e071"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
